Firestore/Source/Local/FSTDataCache.mm

```diff
@@ -0,0 +1,261 @@
+
+#include <memory>
+
+#import "Firestore/Source/Local/FSTDataCache.h"
+#import "Firestore/Source/Local/FSTMutationQueue.h"
+#import "Firestore/Source/Local/FSTQueryCache.h"
+#import "Firestore/Source/Local/FSTRemoteDocumentCache.h"
+#import "Firestore/Source/Core/FSTSnapshotVersion.h"
+#import "Firestore/Source/Core/FSTQuery.h"
+#import "Firestore/Source/Remote/FSTRemoteEvent.h"
+#import "Firestore/Source/Core/FSTListenSequence.h"
+#import "Firestore/Source/Util/FSTAssert.h"
+#import "Firestore/Source/Local/FSTRemoteDocumentChangeBuffer.h"
+#import "Firestore/Source/Local/FSTPersistence.h"
+
+#include "Firestore/core/src/firebase/firestore/core/target_id_generator.h"
+
+using firebase::firestore::core::TargetIdGenerator;
+
+NS_ASSUME_NONNULL_BEGIN
+
+typedef struct {
+    id<FSTMutationQueue> mutationQueue;
+    id<FSTQueryCache> queryCache;
+    id<FSTRemoteDocumentCache> documentCache;
+} table_drivers;
+
+@protocol FSTPersistenceCleanupDelegate
+
+- (void)handlePotentiallyOrphanedDocument:(FSTDocumentKey *)key
+                           sequenceNumber:(FSTListenSequenceNumber)sequenceNumber
+                             exemptTarget:(FSTTargetID)targetID
+                                  drivers:(table_drivers *)drivers
+                                    group:(FSTWriteGroup *)group;
+
+- (void)targetWasRemoved:(FSTTargetID)targetID
+                 drivers:(table_drivers *)drivers
+                   group:(FSTWriteGroup *)group;
+
+@end
+
+@implementation FSTDataCache {
+  std::unique_ptr<table_drivers> _table_drivers;
+  FSTListenSequence *_listenSequence;
+  id<FSTPersistenceCleanupDelegate> _delegate;
+  TargetIdGenerator _targetIDGenerator;
+}
+
+- (id<FSTMutationQueue>)mutationQueue {
+  return _table_drivers->mutationQueue;
+}
+
+- (void)start {}
+
+- (void)shutdown {
+  // TODO: shut down persistence
+}
+
+- (FSTTargetID)highestTargetID {
+  return [_table_drivers->queryCache highestTargetID];
+}
+
+- (FSTSnapshotVersion *)lastRemoteSnapshotVersion {
+  return [_table_drivers->queryCache lastRemoteSnapshotVersion];
+}
+
+- (FSTQueryData *)queryDataForQuery:(FSTQuery *)query {
+  return [_table_drivers->queryCache queryDataForQuery:query];
+}
+
+- (FSTDocumentKeySet *)documentsForTarget:(FSTTargetID)targetID {
+  return [_table_drivers->queryCache matchingKeysForTargetID:targetID];
+}
+
+- (FSTDocumentDictionary *)documentsMatchingQuery:(FSTQuery *)query {
+  return [_table_drivers->documentCache documentsMatchingQuery:query];
+}
+
+- (nullable FSTMaybeDocument *)documentForKey:(FSTDocumentKey *)documentKey {
+  return [_table_drivers->documentCache entryForKey:documentKey];
+}
+
+- (FSTQueryData *)updateQuery:(FSTTargetID)targetID forChange:(FSTTargetChange *)change group:(FSTWriteGroup *)group {
```

Comment:
The real value that this class is adding is that it adds a higher level of abstraction over the query cache, though it seems just as reasonable to make methods like `updateQuery:forChange:group:` just private on the local store. This separate class isn't any more testable than the local store already is (or isn't).